### ISSUE TITLE
batchInsert not returning all ids

### ISSUE THREAD

**Issue by danab** [knex/knex#1828]
I have not been able to get `knex.batchInsert` to return the ids of the new inserted rows. It does in fact insert them as expected, but I only get the first ID back. My code is roughly as follows.

``` javascript
	return knex.batchInsert('table', table_rows )
		.returning( 'id' ) // I've tried with our without this statement
		.then( function( ) {
			console.log( arguments ); // logs { '0': [ 2516942 ] }, I expect { '0': [ 2516942, 2516943 ] }
		})
```

I can confirm the insert is happening in the database, and that table_rows variable is an array of length 2 (or more).

I'm using a mysql connection with knex@0.12.6

**Comment by oscaroox** [comment]
@danab Mysql only returns a single id after a (batch) insert with [LAST_INSERT_ID()](http://dev.mysql.com/doc/refman/5.6/en/information-functions.html#function_last-insert-id)

**Comment by danab** [comment]
This totally makes sense, and I'm fine closing. But two questions:

1) In regards to Knex, would it be better to have a consistent API regardless of the underlying database? I.e., even if it requires a hack, would be it be better to return all the IDs, which somebody might be expecting if they came from a different database.

2) In regards to mySQL, can you trust that the batch inserts all happened successively? I.e. would it even be possible to implement my first option, or it is possible that a concurrent connection might insert a row in between the batch insert.

**Comment by elhigu** [comment]
As far as I know (we have investigated it quite a bit) there is no robust/performant way to get all id's of inserted rows with mysql... In theory you might be able to use e.g. creation time of each row (at least with postgres all inserted rows will get the exact same timestamp) or to lock table before starting to add rows, when you can trust your id sequence being linear inside the transaction.

One way how you could implement that would be to add some extra column called `batchId` and set  that to some unique value and after the insert you would query all id's of rows added with that `batchId`. But to me thats application specific and pretty intrusive implementation and shouldn't be done by knex by default.

About the question 2. I'm not sure how id sequence works if multiple inserts are done in transaction at the same time when other connection is also adding rows to table... need to try that out.

**Comment by ggregoire** [comment]
According to [this answer](http://stackoverflow.com/a/16592867/652669), 

- if your MySQL instance uses the InnoDB engine,
- if your table is in AUTO_INCREMENT,
- if your `innodb_autoinc_lock_mode` is set to 0 or 1 (the default setting),

then it seems you can deduct the IDs by incrementation.

Example: 

```javascript
const [A_ID] = await db('...').insert([A, B, C]) // e.g. 5
const [B_ID, C_ID] = [B, C].map((_, i) => A_ID + i + 1) // 6, 7
```

Should work with `insert` and `batchInsert`.

Can someone confirm?

**Comment by elhigu** [comment]
@ggregoire Sounds legit if table indeed has auto incrementing primary key column and innodb is used... Usually that is the case I suppose. 

It will be pretty nasty to check that all the constraints are fulfilled, so I would rather give the responsibility of satisfying them to the user. So we wouldn't have to keep map of tables and update their column info every now and then...

Maybe query builder could have some special method like returning for mySql, which returns auto_incremented ids of the made insert like `mySqlReturnIdsOfAutoIncrementingColumn()` ...

That way library user will knows that there are restrictions when he can use it (one has to check documentation of the method). 

Or maybe it could be added as another parameter for `.returning()` if people want to write e.g. `.returning(['id'], { returnAutoIncrementedIdsWithMySql: true})` which will enable returning id's the same way with mysql and postgres and oracle.

**Comment by FrankFang** [comment]
Why does mysql do this?

**Comment by elhigu** [comment]
@FrankFang That you should ask from people who has created mysql database.

**Comment by kibertoad** [comment]
@asngeo PR would be welcome!

**Comment by mercmobily** [comment]
This issue is now being tracked as part of a comprehensive solution in issue #6254 "Unified Solution for Cross-Database RETURNING Inconsistencies", which addresses the broader problem of inconsistent returning behaviour across all database dialects in Knex.

The core problem you've identified - MySQL's `LAST_INSERT_ID()` only returning the first ID from a batch insert - is a fundamental MySQL limitation. Unlike PostgreSQL which has native RETURNING clause support, MySQL can only provide the first auto-incremented ID from batch operations. This isn't a Knex bug but rather MySQL's architectural constraint, as confirmed by maintainer @elhigu's investigation showing there's no robust or performant way to retrieve all IDs from batch inserts in MySQL.

Issue #6254 proposes a unified solution through a new `.returningUnified()` method that would provide consistent behaviour across all databases, clearly indicating what each database can and cannot support. For MySQL, this would honestly return empty arrays for `ids` and `rows` whilst always providing the `firstId`, making the limitations transparent rather than surprising.

I'm marking this issue as a duplicate of #6254, where the comprehensive solution is being developed. Please follow and contribute to the discussion there, as it will address not only this MySQL batch insert limitation but also similar inconsistencies across SQLite, MariaDB, MSSQL, and other database dialects.

**Duplicate of:** #6254
